### PR TITLE
Search: SSS integration guide

### PR DESCRIPTION
**Pull request by stsewd**
I'm documenting the process of our SSS indexing, and how we override the default search.

Note that search for mkdocs is still in beta, and we have 2 ways of indexing content now: from the `search_index.json` file (we are losing some content here like code blocks) and from the html page itself (this works good so far).

I'm also editing the parser to support a more general structure as mentioned in this document based on https://webaim.org/techniques/semanticstructure/.

Closes https://github.com/readthedocs/readthedocs.org/issues/4588

### CODE REVIEW

File: docs/guides/search-indexing.rst [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+   :emphasize-lines: 10-12
+
+   <html>
+      <head>
+         ...
+      <head>
+      <body>
+         <div>
+            ...
+         </div>
```
[the remark this answers is not in the document]
Comment (stsewd) [reply docs/guides/search-indexing.rst:46]: Any tag works, we only start indexing from the main node

File: docs/guides/search-indexing.rst [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+      <h1 id="3">
+         ...
+      </h1>
+      <p>
+         ...
+      </p>
+   </div>
+
+Sections can also be wrapped till two levels, and it's parent can contain the id attribute.
```
[the remark this answers is not in the document]
Comment (stsewd) [reply docs/guides/search-indexing.rst:120]: So, we don't have something like sub-sections, all sections are just treated as sections. If you have nested sections (like the example above). I have improved the examples and wording.

File: docs/guides/search-indexing.rst [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+         <p>
+            Content to be indexed
+         </p>
+         <ul>
+            <li>This is also part of the section</li>
+         </ul>
+
+         <div class="section">
+            <div id="another-section">
+               <h2>
```
[the remark this answers is not in the document]
Comment (stsewd) [reply docs/guides/search-indexing.rst:140]: Sections are allowed to be wrapped and its parent can contain the id attribute, I have changed the wording and examples

**Comment by stsewd** [comment]
> I'm not sure if it's in the scope of this document, but maybe worth to mention what's the difference between main, section and subsection regarding indexing. Is it about weight in results?

We only have sections, there isn't a main section or subsections. The only thing special is that the title from the first section will be taken as the title of the page (this is mentioned in the doc)

> Also, we mention semantic structure where section, article and main are used. However, we do no provide examples on those. Does it work the same way, like <article role="main"> ...?

We don't mention semantic structure, we mention ARIA roles, any tag can contain a `main` role.

> How does our parser behaves with HTML4 and HTML5 writers in Sphinx?

Yes, since we parse based on ARIA roles, not on html tags (well, we fallback to the main tag if there isn't a main role)

File: docs/guides/search-indexing.rst [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+         for (var i = 0; i < results.length; i += 1) {
+            var result = process_result(results[i]);
+            append_result(result, search_results);
+         }
+      } else {
+         original_search();
+      }
+   }
+
+   function init_override = function () {
```
Comment (saadmk11) [inline docs/guides/search-indexing.rst:277]:
```suggestion
   var init_override = function () {
```

Comment (stsewd) [reply docs/guides/search-indexing.rst]: oops, thanks!

File: docs/guides/search-indexing.rst [hunk 5]
```diff
@@ -0,0 +1,322 @@
+Better Integration with Server Side Search
+==========================================
+
+Read the Docs provides :doc:`server side search (SSS) </server-side-search>`
+in replace of the default search engine of your site.
+To accomplish this, Read the Docs parses the content directly from your HTML pages [*]_.
+
+If you are the author of a theme or a static site generator you can read the next document,
```
Comment (ericholscher) [inline docs/guides/search-indexing.rst:8]:
```suggestion
If you are the author of a theme or a static site generator you can read this document,
```

File: docs/guides/search-indexing.rst [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+and follow some conventions in order to improve the integration of SSS with your theme/site.
```
Comment (ericholscher) [inline docs/guides/search-indexing.rst:9]:
```suggestion
and follow some conventions to improve the integration of search with your theme/site.
```

File: docs/guides/platform.rst [hunk 7]
```diff
@@ -11,12 +11,13 @@ These guides will help you customize or tune aspects of Read the Docs.
     build-using-too-many-resources
     canonical
     conda
+    embedding-content
     environment-variables
     feature-flags
     google-analytics
     hiding-a-version
+    search-indexing
```
Comment (ericholscher) [inline docs/guides/platform.rst:19]: I'm not sure this is a guide, or something more technical and deep. It feels like guides are user-focused and this is definitely developer focused, so probably belongs somewhere else in our docs?

File: docs/guides/search-indexing.rst [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+   Following the ARIA_ conventions will also improve the accessibility of your site.
+   See also https://webaim.org/techniques/semanticstructure/.
+
+.. _ARIA: https://www.w3.org/TR/wai-aria/
+
+Main content node
+~~~~~~~~~~~~~~~~~
+
+The main content node should have a main role (or a ``main`` tag), and should exists only one per page.
```
Comment (ericholscher) [inline docs/guides/search-indexing.rst:33]:
```suggestion
The main content node should have a main role (or a ``main`` tag), and there should only be one per page.
```

File: docs/guides/search-indexing.rst [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+      <h1 id="neigbor-section">
+         This section is neighbor of "section-title"
+      </h1>
+      <p>
+         ...
+      </p>
+   </div>
+
+Sections can also be wrapped till two levels (and have nested sections),
```
Comment (ericholscher) [inline docs/guides/search-indexing.rst:125]: I don't understand this line "wrapped till two levels"

File: docs/guides/search-indexing.rst [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+The js part of this plugin is included in the `templates/search/main.js`_ file,
+which subscribes to the ``keyup`` event of the ``#mkdocs-search-query`` element
+to call the ``doSearch`` function (available on MkDocs >= 1.x) on every key press.
+
+Read the Docs overrides the ``initSearch`` and ``doSearch`` functions
+to subscribe to the ``keyup`` event of the ``#mkdocs-search-query`` element,
+and puts the results into the ``#mkdocs-search-results`` element.
+A simplified example looks like this:
+
+.. code-block:: js
```
Comment (ericholscher) [inline docs/guides/search-indexing.rst:262]: This is a really complex code example in our docs, and seems likely to get out of date. It's super useful though, so I think it's fine, I just worry about it being super specific vs. something a bit more generic.

File: docs/guides/platform.rst (hunk 7 above)
Comment (stsewd) [reply docs/guides/platform.rst:19]: yeah, I agree this isn't user focused. What about https://github.com/readthedocs/readthedocs.org/blob/sss-integration-guide/docs/development/index.rst?

File: docs/guides/search-indexing.rst (hunk 9 above)
Comment (stsewd) [reply docs/guides/search-indexing.rst]: What about

> Sections can be inside till two nested tags (and have nested sections),

File: docs/guides/platform.rst (hunk 7 above)
Comment (stsewd) [reply docs/guides/platform.rst]: I just moved it there

**Pull request merged**